std_execution.bs

```diff
@@ -3022,3 +3039,152 @@ are all well-formed.
 
     2. Otherwise, `execution::start_detached(execution::then(execution::schedule(sch), f))`.
+
+## Coroutine utilities <b>[execution.coro_utils]</b> ## {#spec-execution.coro_utils}
+
+### `execution::as_awaitable` <b>[execution.coro_utils.as_awaitable]</b> ### {#spec-execution.coro_utils.as_awaitable}
+
+1. `as_awaitable` is used to wrap a sender into an object that is awaitable within a particular coroutine. A sender `s` of type `S` can be made awaitable if and only if:
+
+    - `S` models the concept `typed_sender`, and
+
+    - It passes zero or one values of a unique type to `execution::set_value`, and
+
+    - It passes only objects of type `exception_ptr` to `exception::set_error`, and
+
+    - The promise type of the awaiting coroutine implements a function called `unhandled_done` that takes no arguments and returns a `coroutine_handle<>`.
+
+    <pre highlight="c++">
+    // exposition only
+    template&lt;class S>
+      using <i>single-sender-value-type</i> = <i>see below</i>;
+
+    // exposition only
+    template&lt;class S>
+      concept <i>single-typed-sender</i> =
+        typed_sender&lt;S> &&
+        requires { typename <i>single-sender-value-type</i>&lt;S>; };
+
+    template&lt;<i>single-typed-sender</i> Sender, class Promise>
+      <i>unspecified</i> as_awaitable(Sender&& s, Promise& p);
+    </pre>
+
+2. Alias template <i>single-sender-value-type</i> is defined as follows:
+
+    - If `sender_traits<remove_cvref_t<S>>::value_types<Tuple, Variant>` would have the form `Variant<Tuple<T>>`, then <code><i>single-sender-value-type</i>&lt;S></code> is an alias for type `T`.
+
+    - Otherwise, if `sender_traits<remove_cvref_t<S>>::value_types<Tuple, Variant>` would have the form `Variant<Tuple<>>` or `Variant<>`, then <code><i>single-sender-value-type</i>&lt;S></code> is an alias for type `void`.
+
+    - Otherwise, <code><i>single-sender-value-type</i>&lt;S</code> is ill-formed.
+
+3. For some expressions `s` and `p` such that `decltype((s))` is `S` and `decltype((p))` is `P&`, `as_awaitable(s, p)` returns <code><i>sender-awaitable</i>{s, p}</code>, where  <code><i>sender-awaitable</i></code> is an unspecified class type equivalant to the following:
+
+    <pre highlight="c++">
+    class <i>sender-awaitable</i> {
+      struct unit {}; // exposition only
+      using value_t = <i>single-sender-value-type</i>&lt;S>;
+      using result_t = conditional_t&lt;is_void_v&lt;value_t>, unit, value_t>;
+      struct <i>awaitable-receiver</i>; // exposition only
+
+      variant&lt;monostate, result_t, exception_ptr> result_{}; // exposition only
+      connect_result_t&lt;S, <i>awaitable-receiver</i>> state_;  // exposition only
+
+     public:
+      <i>sender-awaitable</i>(S&& s, P& p);
+      bool await_ready() const noexcept { return false; }
+      void await_suspend(coro::coroutine_handle&lt;>) noexcept { start(state_); }
+      value_t await_resume();
```

Comment:
If `value_types<Tuple, Variant>` has form `Variant<>` then maybe this should be marked `[[nodiscard]]`?

And if `error_types<Variant>` has form `Variant<>` then maybe this should be marked `noexcept`?

Reply:
Deferring this.

Reply:
> If `value_types<Tuple, Variant>` has form `Variant<>` then maybe this should be marked `[[nodiscard]]`?

@lewissbaker, wouldn't that mean that this awaitable will never produce a value, and so `await_resume` will never be called? What would be the point of declaring it `[[nodiscard]]`?

> And if `error_types<Variant>` has form `Variant<>` then maybe this should be marked `noexcept`?

Filed as #213